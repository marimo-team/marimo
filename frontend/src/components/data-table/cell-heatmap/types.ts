/* Copyright 2024 Marimo. All rights reserved. */
/* eslint-disable @typescript-eslint/no-empty-interface */
import type { RowData, Updater } from "@tanstack/react-table";

export interface CellHeatmapTableState {
  columnHeatmap: Record<string, boolean>;
  cachedMinValue?: number | null;
  cachedMaxValue?: number | null;
}

export interface CellHeatmapOptions {
  enableCellHeatmap?: boolean;
  onGlobalHeatmapChange?: (updater: Updater<boolean>) => void;
  onColumnHeatmapChange?: (updater: Updater<Record<string, boolean>>) => void;
}

export interface CellHeatmapState {
  global: boolean;
  columns: Record<string, boolean>;
}

// Use declaration merging to add our new feature APIs
declare module "@tanstack/react-table" {
  interface TableState extends CellHeatmapTableState {}
  interface InitialTableState extends CellHeatmapTableState {}

  interface TableOptionsResolved<TData extends RowData>
    extends CellHeatmapOptions {}

  interface Table<TData extends RowData> {
    getGlobalHeatmap?: () => boolean;
    toggleGlobalHeatmap?: (value?: boolean) => void;
  }

  interface Column<TData extends RowData> {
    getCellHeatmapColor?: (cellValue: unknown) => string;
    toggleColumnHeatmap?: (value?: boolean) => void;
    getIsColumnHeatmapEnabled?: () => boolean;
  }
}
