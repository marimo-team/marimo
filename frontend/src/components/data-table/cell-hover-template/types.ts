/* Copyright 2024 Marimo. All rights reserved. */
/* eslint-disable @typescript-eslint/no-empty-interface */

export interface CellHoverTemplateTableState {
  cellHoverTemplate: string | null;
}

// Use declaration merging to add our new feature APIs
declare module "@tanstack/react-table" {
  interface TableState extends CellHoverTemplateTableState {}
}
