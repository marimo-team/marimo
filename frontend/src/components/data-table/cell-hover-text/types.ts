/* Copyright 2024 Marimo. All rights reserved. */
/* eslint-disable @typescript-eslint/no-empty-interface */

import type { RowData } from "@tanstack/react-table";

export type CellHoverTextState = Record<string, Record<string, string | null>>;

export interface CellHoverTextTableState {
  cellHoverTexts: CellHoverTextState;
}

export interface CellHoverTextCell {
  /**
   * Returns precomputed hover text for the cell, if any.
   */
  getHoverTitle?: () => string | undefined | null;
}

// Use declaration merging to add our new feature APIs
declare module "@tanstack/react-table" {
  interface TableState extends CellHoverTextTableState {}

  interface Cell<TData extends RowData, TValue> extends CellHoverTextCell {}
}
