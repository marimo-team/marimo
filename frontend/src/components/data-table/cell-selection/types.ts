/* Copyright 2024 Marimo. All rights reserved. */
/* eslint-disable @typescript-eslint/no-empty-interface */
import type { OnChangeFn, RowData, Updater } from "@tanstack/react-table";

export interface CellSelectionItem {
  rowId: string;
  columnName: string;
}
export type CellSelectionState = CellSelectionItem[];
export interface CellSelectionTableState {
  cellSelection: CellSelectionState;
}

export interface CellSelectionOptions {
  enableCellSelection?: boolean;
  enableMultiCellSelection?: boolean;
  onCellSelectionChange?: OnChangeFn<CellSelectionState>; // | CellSelectionItem ???
}

export interface CellSelectionCell {
  /**
   * Returns whether or not the cell is selected.
   */
  getIsSelected?: () => boolean;

  /**
   * Selects/deselects the cell.
   */
  toggleSelected?: (value?: boolean) => void;
}

// These are additional properties add to the table instance in createTable
export interface CellSelectionInstance<TData extends RowData> {
  setCellSelection?: (updater: Updater<CellSelectionState>) => void;

  resetCellSelection?(defaultValue?: CellSelectionState): void;
}

// Use declaration merging to add our new feature APIs
declare module "@tanstack/react-table" {
  interface TableState extends CellSelectionTableState {}

  interface TableOptionsResolved<TData extends RowData>
    extends CellSelectionOptions {}

  interface Table<TData extends RowData> extends CellSelectionInstance<TData> {}

  interface Cell<TData extends RowData, TValue> extends CellSelectionCell {}
}
