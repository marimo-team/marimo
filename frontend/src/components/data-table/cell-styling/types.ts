/* Copyright 2024 Marimo. All rights reserved. */
/* eslint-disable @typescript-eslint/no-empty-interface */
import type { RowData } from "@tanstack/react-table";

export type CellStyleState = Record<
  string,
  Record<string, React.CSSProperties>
>;

export interface CellStylingTableState {
  cellStyling: CellStyleState | undefined | null;
}

export interface CellStylingCell {
  /**
   * Returns additional styling for the cell.
   */
  getUserStyling?: () => React.CSSProperties;
}

// Use declaration merging to add our new feature APIs
declare module "@tanstack/react-table" {
  interface TableState extends CellStylingTableState {}

  interface Cell<TData extends RowData, TValue> extends CellStylingCell {}
}
