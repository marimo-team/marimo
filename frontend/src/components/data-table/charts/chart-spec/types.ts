/* Copyright 2024 Marimo. All rights reserved. */

import type { DataType } from "@/core/kernel/messages";
import { ChartType, type SelectableDataType } from "../types";
import type { StandardType } from "vega-lite/build/src/type";
import { logNever } from "@/utils/assertNever";
import type { Mark } from "@/plugins/impl/vega/types";

export function convertDataTypeToVega(
  dataType: DataType | SelectableDataType,
): StandardType {
  switch (dataType) {
    case "number":
    case "integer":
      return "quantitative";
    case "string":
    case "boolean":
    case "unknown":
      return "nominal";
    case "date":
    case "datetime":
    case "time":
    case "temporal":
      return "temporal";
    default:
      logNever(dataType);
      return "nominal";
  }
}

export function convertDataTypeToSelectable(
  type: DataType,
): SelectableDataType {
  switch (type) {
    case "number":
    case "integer":
      return "number";
    case "string":
    case "boolean":
    case "unknown":
      return "string";
    case "date":
    case "datetime":
    case "time":
      return "temporal";
    default:
      logNever(type);
      return "string";
  }
}

export function convertChartTypeToMark(chartType: ChartType): Mark {
  switch (chartType) {
    case ChartType.PIE:
      return "arc";
    case ChartType.SCATTER:
      return "point";
    case ChartType.HEATMAP:
      return "rect";
    default:
      return chartType;
  }
}
