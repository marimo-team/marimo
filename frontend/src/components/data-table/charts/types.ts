/* Copyright 2024 Marimo. All rights reserved. */

/**
 * Similar to VegaLite's ScaleType, https://vega.github.io/vega-lite/docs/scale.html#type
 */
export const SELECTABLE_DATA_TYPES = ["number", "string", "temporal"] as const;
export type SelectableDataType = (typeof SELECTABLE_DATA_TYPES)[number];

/**
 * Used for adding data types in Altair encoding
 */
export const DATA_TYPE_LETTERS: Record<SelectableDataType, string> = {
  number: "Q",
  string: "N",
  temporal: "T",
} as const;
export type DataTypeLetter =
  (typeof DATA_TYPE_LETTERS)[keyof typeof DATA_TYPE_LETTERS];

/**
 * Similar to VegaLite's TimeUnit, https://vega.github.io/vega-lite/docs/timeunit.html
 */
export const SINGLE_TIME_UNITS = [
  // Individual units
  "year", // Gregorian calendar years.
  "quarter", // Three-month intervals, starting in one of January, April, July, and October.
  "month", // Calendar months (January, February, etc.).
  "date", // Calendar day of the month (January 1, January 2, etc.).
  "week", // Sunday-based weeks. Days before the first Sunday of the year are considered to be in week 0, the first Sunday of the year is the start of week 1, the second Sunday week 2, etc..
  "day", // Day of the week (Sunday, Monday, etc.).
  "dayofyear", // Day of the year (1, 2, â€¦, 365, etc.).
  "hours", // Hours of the day (12:00am, 1:00am, etc.).
  "minutes", // Minutes in an hour (12:00, 12:01, etc.).
  "seconds", // Seconds in a minute (12:00:00, 12:00:01, etc.).
  "milliseconds", // Milliseconds in a second.
] as const;
// Common combinations of the above
export const COMBINED_TIME_UNITS = [
  "yearmonth",
  "yearmonthdate",
  "monthdate",
] as const;
export const TIME_UNITS = [
  ...SINGLE_TIME_UNITS,
  ...COMBINED_TIME_UNITS,
] as const;
export type TimeUnit = (typeof TIME_UNITS)[number];

// Time units that are not selectable options but are used for tooltips
export const TIME_UNIT_TOOLTIPS = [
  ...TIME_UNITS,
  "yearmonthdatehoursminutesseconds",
  "yearmonthdate",
  "hoursminutesseconds",
] as const;
export type TimeUnitTooltip = (typeof TIME_UNIT_TOOLTIPS)[number];

/**
 * Similar to VegaLite's SortOrder, https://vega.github.io/vega-lite/docs/sort.html#order
 */
export const SORT_TYPES = ["ascending", "descending"] as const;
export type SortType = (typeof SORT_TYPES)[number];

export const NONE_AGGREGATION = "none";
export const BIN_AGGREGATION = "bin"; // We use this not to aggregate, but to bin

/**
 * Subset of VegaLite's AggregateOp, https://vega.github.io/vega-lite/docs/aggregate.html#op
 */
export const AGGREGATION_FNS = [
  NONE_AGGREGATION,
  "count",
  "sum",
  "mean",
  "median",
  "min",
  "max",
  "distinct",
  "valid",
  "stdev",
  "stdevp",
  "variance",
  "variancep",
  BIN_AGGREGATION,
] as const;
export type AggregationFn = (typeof AGGREGATION_FNS)[number];

/*
 * Subset of AGGREGATION_FNS that are valid for string data types
 */
export const STRING_AGGREGATION_FNS: AggregationFn[] = [
  "none",
  "count",
  "distinct",
  "valid",
];

/**
 * Subset of VegaLite's MarkType, https://vega.github.io/vega-lite/docs/mark.html#types
 */
export const ChartType = {
  LINE: "line",
  BAR: "bar",
  PIE: "pie",
  SCATTER: "scatter",
  HEATMAP: "heatmap",
  AREA: "area",
} as const;
export type ChartType = (typeof ChartType)[keyof typeof ChartType];
export const CHART_TYPES = Object.values(ChartType);
