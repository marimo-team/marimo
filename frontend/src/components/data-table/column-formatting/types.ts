/* Copyright 2024 Marimo. All rights reserved. */
/* eslint-disable @typescript-eslint/no-empty-interface */
import type { DataType } from "@/core/kernel/messages";
import type { OnChangeFn, RowData } from "@tanstack/react-table";

// define all format options
export const formatOptions = {
  date: ["Date", "Datetime", "Time"],
  integer: ["Auto", "Percent", "Scientific", "Integer"],
  number: ["Auto", "Percent", "Scientific", "Integer"],
  string: ["Uppercase", "Lowercase", "Capitalize", "Title"],
  boolean: ["Yes/No", "On/Off"],
  unknown: [],
} as const satisfies Record<DataType, string[]>;

// define types for format options
export type FormatOptions = (typeof formatOptions)[keyof typeof formatOptions];
export type FormatOption = FormatOptions[number];

// define types for column formatting's custom state
export type ColumnFormattingState = Record<string, FormatOption | undefined>;
export interface ColumnFormattingTableState {
  columnFormatting: ColumnFormattingState;
}

// define types for column formatting's table options
export interface ColumnFormattingOptions {
  enableColumnFormatting?: boolean;
  onColumnFormattingChange?: OnChangeFn<ColumnFormattingState>;
}

// define types for column formatting's table APIs
export interface ColumnFormattingInstance {
  setColumnFormatting: (value?: FormatOption) => void;
  getColumnFormatting?: () => FormatOption | undefined;
  getCanFormat?: () => boolean;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  applyColumnFormatting: (value: any) => any;
}

// Use declaration merging to add APIs and state types
// to TanStack Table's existing types.
declare module "@tanstack/react-table" {
  //merge column formatting's state with the existing table state
  interface TableState extends ColumnFormattingTableState {}
  //merge column formatting's options with the existing table options
  interface TableOptionsResolved<TData extends RowData>
    extends ColumnFormattingOptions {}
  //merge column formatting's instance APIs with the existing table instance APIs
  interface Column<TData extends RowData> extends ColumnFormattingInstance {}
}
