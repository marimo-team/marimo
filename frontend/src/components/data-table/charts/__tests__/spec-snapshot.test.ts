/* Copyright 2024 Marimo. All rights reserved. */

import { describe, it, expect } from "vitest";
import {
  createSpecWithoutData,
  augmentSpecWithData,
  X_AXIS_REQUIRED,
  Y_AXIS_REQUIRED,
} from "../chart-spec/spec";
import type { ChartSchemaType } from "../schemas";
import { NONE_AGGREGATION, ChartType } from "../types";
import type { TopLevelSpec } from "vega-lite/build/src/spec";

describe("create vega spec", () => {
  // Sample data for testing
  const sampleData = [
    { category: "A", value: 10, group: "Group 1" },
    { category: "B", value: 20, group: "Group 1" },
    { category: "C", value: 15, group: "Group 1" },
    { category: "A", value: 5, group: "Group 2" },
    { category: "B", value: 10, group: "Group 2" },
    { category: "C", value: 25, group: "Group 2" },
  ];

  // Helper function to create basic form values
  const createBasicFormValues = (): ChartSchemaType => ({
    general: {
      title: "Test Chart",
      xColumn: {
        field: "category",
        type: "string" as const,
      },
      yColumn: {
        field: "value",
        type: "number" as const,
        aggregate: NONE_AGGREGATION,
      },
    },
  });

  it("should create and augment a spec", () => {
    const spec = createSpecWithoutData(
      ChartType.BAR,
      createBasicFormValues(),
      "light",
      400,
      300,
    );
    expect(spec).toMatchSnapshot();
    expect(typeof spec !== "string").toBe(true); // Not error message
    expect((spec as TopLevelSpec).data).toEqual({ values: [] });

    // Augment the spec with data
    const augmentedSpec = augmentSpecWithData(spec as TopLevelSpec, sampleData);
    expect(augmentedSpec.data).toEqual({ values: sampleData });
  });

  it("should return an error message if the spec is invalid", () => {
    const formValues = createBasicFormValues();
    formValues.general!.xColumn!.field = undefined;

    const spec = createSpecWithoutData(
      ChartType.BAR,
      formValues,
      "light",
      400,
      300,
    );
    expect(spec).toEqual(X_AXIS_REQUIRED);

    // Undefined yColumn
    const formValues2 = createBasicFormValues();
    formValues2.general!.yColumn!.field = undefined;
    const spec2 = createSpecWithoutData(
      ChartType.BAR,
      formValues2,
      "light",
      400,
      300,
    );
    expect(spec2).toEqual(Y_AXIS_REQUIRED);
  });
});
