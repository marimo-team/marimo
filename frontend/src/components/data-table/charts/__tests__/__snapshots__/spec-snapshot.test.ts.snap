// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createVegaSpec > Bar Chart > should create a horizontal bar chart 1`] = `
{
  "$schema": "https://vega.github.io/schema/vega-lite/v5.json",
  "background": "white",
  "data": {
    "values": [
      {
        "category": "A",
        "group": "Group 1",
        "value": 10,
      },
      {
        "category": "B",
        "group": "Group 1",
        "value": 20,
      },
      {
        "category": "C",
        "group": "Group 1",
        "value": 15,
      },
      {
        "category": "A",
        "group": "Group 2",
        "value": 5,
      },
      {
        "category": "B",
        "group": "Group 2",
        "value": 10,
      },
      {
        "category": "C",
        "group": "Group 2",
        "value": 25,
      },
    ],
  },
  "encoding": {
    "x": {
      "field": "value",
      "type": "nominal",
    },
    "y": {
      "field": "category",
      "type": "nominal",
    },
  },
  "height": 300,
  "mark": {
    "type": "bar",
  },
  "resolve": {
    "axis": {
      "x": "independent",
      "y": "independent",
    },
  },
  "title": "Test Chart",
  "width": 400,
}
`;

exports[`createVegaSpec > Bar Chart > should create a stacked bar chart with grouping 1`] = `
{
  "$schema": "https://vega.github.io/schema/vega-lite/v5.json",
  "background": "white",
  "data": {
    "values": [
      {
        "category": "A",
        "group": "Group 1",
        "value": 10,
      },
      {
        "category": "B",
        "group": "Group 1",
        "value": 20,
      },
      {
        "category": "C",
        "group": "Group 1",
        "value": 15,
      },
      {
        "category": "A",
        "group": "Group 2",
        "value": 5,
      },
      {
        "category": "B",
        "group": "Group 2",
        "value": 10,
      },
      {
        "category": "C",
        "group": "Group 2",
        "value": 25,
      },
    ],
  },
  "encoding": {
    "color": {
      "field": "group",
      "scale": {},
      "type": "nominal",
    },
    "x": {
      "field": "category",
      "type": "nominal",
    },
    "y": {
      "field": "value",
      "stack": true,
      "type": "nominal",
    },
  },
  "height": 300,
  "mark": {
    "type": "bar",
  },
  "resolve": {
    "axis": {
      "x": "independent",
      "y": "independent",
    },
  },
  "title": "Test Chart",
  "width": 400,
}
`;

exports[`createVegaSpec > Edge cases > should handle missing xColumn field 1`] = `"X-axis column is required"`;

exports[`createVegaSpec > Edge cases > should handle missing yColumn field 1`] = `"Y-axis column is required"`;

exports[`createVegaSpec > Line Chart > should create a basic line chart spec 1`] = `
{
  "$schema": "https://vega.github.io/schema/vega-lite/v5.json",
  "background": "white",
  "data": {
    "values": [
      {
        "category": "A",
        "group": "Group 1",
        "value": 10,
      },
      {
        "category": "B",
        "group": "Group 1",
        "value": 20,
      },
      {
        "category": "C",
        "group": "Group 1",
        "value": 15,
      },
      {
        "category": "A",
        "group": "Group 2",
        "value": 5,
      },
      {
        "category": "B",
        "group": "Group 2",
        "value": 10,
      },
      {
        "category": "C",
        "group": "Group 2",
        "value": 25,
      },
    ],
  },
  "encoding": {
    "x": {
      "field": "category",
      "type": "nominal",
    },
    "y": {
      "field": "value",
      "type": "nominal",
    },
  },
  "height": 300,
  "mark": {
    "type": "line",
  },
  "resolve": {
    "axis": {
      "x": "independent",
      "y": "independent",
    },
  },
  "title": "Test Chart",
  "width": 400,
}
`;

exports[`createVegaSpec > Pie Chart > should create a pie chart with tooltips 1`] = `"Color by column is required"`;

exports[`createVegaSpec > Scatter Chart > should create a scatter chart with grouping 1`] = `
{
  "$schema": "https://vega.github.io/schema/vega-lite/v5.json",
  "background": "white",
  "data": {
    "values": [
      {
        "category": "A",
        "group": "Group 1",
        "value": 10,
      },
      {
        "category": "B",
        "group": "Group 1",
        "value": 20,
      },
      {
        "category": "C",
        "group": "Group 1",
        "value": 15,
      },
      {
        "category": "A",
        "group": "Group 2",
        "value": 5,
      },
      {
        "category": "B",
        "group": "Group 2",
        "value": 10,
      },
      {
        "category": "C",
        "group": "Group 2",
        "value": 25,
      },
    ],
  },
  "encoding": {
    "color": {
      "field": "group",
      "scale": {},
      "type": "nominal",
    },
    "x": {
      "field": "category",
      "type": "nominal",
    },
    "y": {
      "field": "value",
      "type": "nominal",
    },
  },
  "height": 300,
  "mark": {
    "type": "point",
  },
  "resolve": {
    "axis": {
      "x": "independent",
      "y": "independent",
    },
  },
  "title": "Test Chart",
  "width": 400,
}
`;

exports[`createVegaSpec > Theme variations > should create a chart with dark theme 1`] = `
{
  "$schema": "https://vega.github.io/schema/vega-lite/v5.json",
  "background": "dark",
  "data": {
    "values": [
      {
        "category": "A",
        "group": "Group 1",
        "value": 10,
      },
      {
        "category": "B",
        "group": "Group 1",
        "value": 20,
      },
      {
        "category": "C",
        "group": "Group 1",
        "value": 15,
      },
      {
        "category": "A",
        "group": "Group 2",
        "value": 5,
      },
      {
        "category": "B",
        "group": "Group 2",
        "value": 10,
      },
      {
        "category": "C",
        "group": "Group 2",
        "value": 25,
      },
    ],
  },
  "encoding": {
    "x": {
      "field": "category",
      "type": "nominal",
    },
    "y": {
      "field": "value",
      "type": "nominal",
    },
  },
  "height": 300,
  "mark": {
    "type": "bar",
  },
  "resolve": {
    "axis": {
      "x": "independent",
      "y": "independent",
    },
  },
  "title": "Test Chart",
  "width": 400,
}
`;

exports[`createVegaSpec > should create a bar chart with binning 1`] = `
{
  "$schema": "https://vega.github.io/schema/vega-lite/v5.json",
  "background": "white",
  "data": {
    "values": [
      {
        "category": "A",
        "group": "Group 1",
        "value": 10,
      },
      {
        "category": "B",
        "group": "Group 1",
        "value": 20,
      },
      {
        "category": "C",
        "group": "Group 1",
        "value": 15,
      },
      {
        "category": "A",
        "group": "Group 2",
        "value": 5,
      },
      {
        "category": "B",
        "group": "Group 2",
        "value": 10,
      },
      {
        "category": "C",
        "group": "Group 2",
        "value": 25,
      },
    ],
  },
  "encoding": {
    "x": {
      "field": "category",
      "type": "nominal",
    },
    "y": {
      "field": "value",
      "type": "nominal",
    },
  },
  "height": 300,
  "mark": {
    "type": "bar",
  },
  "resolve": {
    "axis": {
      "x": "independent",
      "y": "independent",
    },
  },
  "title": "Test Chart",
  "width": 400,
}
`;
